did_peer/src/key/verification_method.rs

```diff
@@ -0,0 +1,228 @@
+        SupportedKeyType::X25519 => VerificationMethodType::X25519KeyAgreementKey2020,
+        SupportedKeyType::P256 => VerificationMethodType::JsonWebKey2020,
+        SupportedKeyType::P384 => VerificationMethodType::JsonWebKey2020,
+        SupportedKeyType::P521 => VerificationMethodType::JsonWebKey2020,
+        SupportedKeyType::Bls12381g1g2 => {
+            let g1_key = Key::new(key.key()[..48].to_vec(), SupportedKeyType::Bls12381g1)?;
+            let g2_key = Key::new(key.key()[48..].to_vec(), SupportedKeyType::Bls12381g2)?;
+            return Ok(build_verification_methods_from_bls_multikey(
+                g1_key,
+                g2_key,
+                id,
+                did.to_owned(),
+                public_key_encoding,
+            ));
+        }
+    };
+    Ok(build_verification_methods_from_type_and_key(
+        vm_type,
+        key,
+        id,
+        did.to_owned(),
+        public_key_encoding,
+    ))
+}
+
+pub fn get_key_by_verification_method(vm: &VerificationMethod) -> Result<Key, DidPeerError> {
+    let key_type = match vm.verification_method_type() {
+        VerificationMethodType::Ed25519VerificationKey2018 | VerificationMethodType::Ed25519VerificationKey2020 => {
+            SupportedKeyType::Ed25519
+        }
+        VerificationMethodType::Bls12381G1Key2020 => SupportedKeyType::Bls12381g1,
+        VerificationMethodType::Bls12381G2Key2020 => SupportedKeyType::Bls12381g2,
+        VerificationMethodType::X25519KeyAgreementKey2019 | VerificationMethodType::X25519KeyAgreementKey2020 => {
+            SupportedKeyType::X25519
+        }
+        t @ _ => return Err(DidPeerError::UnsupportedVerificationMethodType(t.to_owned())),
+    };
+    Key::new(vm.public_key().key_decoded()?, key_type)
+}
+
+fn build_verification_methods_from_type_and_key(
+    vm_type: VerificationMethodType,
+    key: &Key,
+    id: DidUrl,
+    did: Did,
+    public_key_encoding: &PublicKeyEncoding,
+) -> Vec<VerificationMethod> {
+    vec![VerificationMethod::builder(id, did, vm_type)]
+        .iter()
+        .map(|builder| add_public_key_to_builder(builder.to_owned(), key, public_key_encoding))
+        .collect::<Vec<VerificationMethod>>()
+}
+
+fn build_verification_methods_from_bls_multikey(
+    g1_key: Key,
+    g2_key: Key,
+    id: DidUrl,
+    did: Did,
+    public_key_encoding: &PublicKeyEncoding,
+) -> Vec<VerificationMethod> {
+    vec![
+        VerificationMethod::builder(id.to_owned(), did.to_owned(), VerificationMethodType::Bls12381G1Key2020),
+        VerificationMethod::builder(id, did, VerificationMethodType::Bls12381G2Key2020),
+    ]
+    .iter()
+    .enumerate()
+    .map(|(index, builder)| {
+        let key = match index {
+            0 => &g1_key,
+            1 => &g2_key,
+            _ => unreachable!(),
+        };
+        add_public_key_to_builder(builder.to_owned(), key, public_key_encoding)
+    })
+    .collect::<Vec<VerificationMethod>>()
```

Comment:
What do you think about first creating the individual verification methods and constructing the `Vec` *after*? This avoids constructing a 2 elements long `Vec` to just iterate over and consume, to create another `Vec`.

Reply:
I have no issue with that, sounds reasonable.